### ISSUE TITLE
Webfinger account search issues

### ISSUE THREAD

**Issue by mariusor** [mastodon/mastodon#17133]
### Steps to reproduce the problem

Hello, I'm developing a web application that uses ActivityPub but I'm having an issue with account search in Mastodon.

First I need to provide some details about my application.  
It is formed of two independent services:

* A web frontend that users interact with and that implements webfinger for compatibility with Mastodon's user discovery (eg, https://littr.me)
* An independent ActivityPub service (which will probably be on a different domain than the frontend) which doesn't have any webfinger endpoints and which has no way of interacting with, except plain ActivityPub requests. (eg, https://federated.id)

When I am on Mastodon and I search for a user `@marius@littr.me` (on domain littr.me which has webfinger support) the reply is a proper webfinger response (as you can see below), however Mastodon takes the domain of the `self` activitypub ref and redoes the search (which now fails, because federated.id does not have a webfinger end-point).

```
$ curl -k https://littr.me/.well-known/webfinger?resource=acct:marius@littr.me
HTTP/2 200 
date: Tue, 14 Dec 2021 12:16:47 GMT
content-type: application/jrd+json

{
    "subject": "acct:marius@littr.me",
    "aliases": [ "https://littr.me/~marius" ],
    "links": [
        {
            "rel": "self",
            "type": "application/activity+json",
            "href": "https://federated.id/actors/8b740680-ccb4-4265-82e7-4ac2ca402750"
        },
        {
            "rel": "https://webfinger.net/rel/profile-page",
            "type": "text/html",
            "href": "https://littr.me/~marius"
        }
    ]
}
```

### Expected behaviour

Mastodon uses the first reply to infer the actor corresponding to the searched user.

### Actual behaviour

Mastodon does another webfinger account search on the wrong domain.

### Specifications

I would assume the browser and OS are irrelevant: but I'm using Firefox Developer Edition (96.0b4 64bit) on Archlinux.

The Mastodon version I encountered this behaviour on was v3.4.4.

Please let me know if there's any more information I could provide.

**Comment by mariusor** [comment]
Any pointers from existing contributors to where this could be addressed in the code, would be welcome, I can try to come up with a patch.

**Comment by trwnh** [comment]
https://github.com/mastodon/mastodon/blob/main/app/services/resolve_account_service.rb maybe?

**Comment by ClearlyClaire** [comment]
As Mastodon requires *all* actors it interacts with to have a canonical `acct:` URI, it must be able to find that canonical `acct:` from the actor's `id`. This is done by fetching the actor, using `preferredName`, and doing one or more webfinger requests. In this case, to be compatible with Mastodon, `https://federated.id/.well-known/webfinger?resource=acct:marius@federated.id` should redirect to (or return the same contents as) `https://littr.me/.well-known/webfinger?resource=acct:marius@littr.me`

**Comment by mariusor** [comment]
Thank you for your reply @ClearlyClaire, but can you expand a bit more on why that is required? 

The resource at the URL that the ID points to should be the canonical representation of the actor, having extra webfinger requests to confirm seems to be just extra work without any benefit. Is there something I'm missing about it?

Like I said federated.id is a service that understands ActivityPub (and just ActivityPub), but it has a different method than webfinger for user discovery. However I would like it very much if it would be possible for users from Mastodon to interact with it.

**Comment by ClearlyClaire** [comment]
> The resource at the URL that the ID points to should be the canonical representation of the actor,

Yes, but unfortunately, the ActivityPub ID of the actor is not the canonical identifier of the actor in Mastodon. For UX and database design reasons that date from before Mastodon embraced ActivityPub, the canonical identifier of the actor in Mastodon is its `acct:` URI.

> having extra webfinger requests to confirm seems to be just extra work without any benefit. Is there something I'm missing about it?

That extra webfinger request is used to confirm this is the *canonical* `acct` URI, to be stored in Mastodon's database. Without it, some features would not be possible (e.g. having a webfinger handle and ActivityPub URI on different domains), and an attacker could simply create aliases for other users, poisoning the database with an “incorrect” handle for a real user.

**Comment by mariusor** [comment]
I suspect that moving towards using an actor's ID as a canonical identifier in Mastodon is not a priority for you guys. Seems like a lot of work for little benefit. Do you know if there is a ticket or desire for this change, even in the long run?

**Comment by trwnh** [comment]
> I suspect that moving towards using an actor's ID as a canonical identifier in Mastodon is not a priority for you guys. Seems like a lot of work for little benefit. Do you know if there is a ticket or desire for this change, even in the long run?

#17030
#10745

**Comment by markcellus** [comment]
@ClearlyClaire when you are searching for the user on mastodon, which instance are you using? When I type my user in the search bar on mastodon.online, it just returns a response from mastodon's `/api/v2/search` endpoint. Even when I search your user there (`@marius@littr.me`), it does the same and shows nothing. How are you able to see the response from the `/.well-known/webfinger` URLs?

**Comment by ClearlyClaire** [comment]
> @ClearlyClaire when you are searching for the user on mastodon, which instance are you using? When I type my user in the search bar on mastodon.online, it just returns a response from mastodon's `/api/v2/search` endpoint. Even when I search your user there (`@marius@littr.me`), it does the same and shows nothing.

I use my personal instance. But querying the REST API is not going to give you a lot of information either way, it is not meant for diagnosis purposes.

> How are you able to see the response from the `/.well-known/webfinger` URLs?

I'm not sure what you're asking. Just accessing them from a browser should do the trick.

For diagnosis purposes, something I sometimes do is use `ResolveAccountService.new.call('foo@bar', suppress_errors: false)` in an instance's Rails console with #15605 merged.

**Comment by markcellus** [comment]
Sorry I wasn't as clear. 

Using `bob@myexample.com` for privacy purposes, `curl`ing `/.well-known/webfinger?resource=acct:bob@my-example.com` directly returns a 200 with the proper contents. But when I search `@bob@myexample.com` in the mastodon search bar on [mastodon.online](https://mastodon.online), it doesn't call any webfinger URLs. The same happens when I try with your user: `@marius@littr.me`.

![image](https://user-images.githubusercontent.com/1446916/147936214-0cea05e6-6063-402b-9f0c-4435e629acd9.png)

How are you testing that your personal mastodon instance is actually calling the webfinger URL? Are you typing in the search bar of your personal instance?

Hope this helps clarify.

**Comment by ClearlyClaire** [comment]
I'm not sure how you can tell it “doesn't call any webfinger URLs” if you do not control `mastodon.online` nor `littr.me` (btw. I'm not related to `littr.me` or `@marius@littr.me`). Webfinger queries are performed server-side by the instance.

Again, the search bar is not made for diagnosis purposes. In fact, Mastodon does not have that kind of diagnosis tools built-in. There is nothing in Mastodon itself for a user to check a webfinger query, or any query at all, has been made. If you are a server admin, you can run your instance with `RAILS_LOG_LEVEL=debug` and it will log which HTTP queries it makes (including webfinger queries), but ultimately, there isn't much debug logging or tools in Mastodon.

What happens when you type a `@foo@bar` handle in the search box is that it checks whether it knows such an account, and whenever it does not, or that account has not been refreshed in a while, it performs a series of queries, starting with a webfinger query, unless the target domain is blocked. If any of these requests fails, it does not return anything.

**Comment by markcellus** [comment]
Oh I see. I'm not using a Mastodon instance. I'm using a basic ActivityPub server and followed all instructions [here](https://blog.joinmastodon.org/2018/06/how-to-implement-a-basic-activitypub-server/). I guess I expected that once I implemented that, my user would be searchable on the internet from any mastodon instance. 

So, it sounds like I can't search users or content on a basic ActivityPub server from the search bar of a mastodon instance or ones that are not on the same domain as the instance. 

Is my understanding of that correct?

**Comment by ClearlyClaire** [comment]
> So, it sounds like I can't search users or content on a basic ActivityPub server from the search bar of a mastodon instance or ones that are not on the same domain as the instance.
> 
> Is my understanding of that correct?

No, your understanding isn't correct. It should work, but something is amiss. However, we currently have no tool built inside Mastodon to debug this.

**Comment by markcellus** [comment]
Ok that's good to know. I'll just run a local instance and debug there like you recommmended. Thanks for the information.

**Comment by nightpool** [comment]
Yeah, it's impossible for us to help you debug without actually being able to see the responses from the server in question. Probably you're missing the content-type on your returned responses though, that seems to be the most common problem for implementors.

**Comment by nightpool** [comment]
@mariusor were you able to understand why canonicalizing the webfinger `acct:` URI in this manner is required for Mastodon? The goal is to make sure that `littr.me` and `federated.id` both acknowledge that `@marius@littr.me` is the canonical URI for this actor, because an implementation that started with `https://federated.id/actors/8b740680-ccb4-4265-82e7-4ac2ca402750` would have no other way of discovering `@marius@littr.me` as the webfinger domain, since it's not referenced anywhere in the ActivityPub actor JSON. Does that make sense? Do you need anything else from us to continue implementing?

**Comment by mariusor** [comment]
@nightpool apologies for the long delay, I actually didn't notice you replied with this very useful information. Thank you very much.

Based on that I would ask if it's still a problem if both the ActivityPub actor and the webfinger response offer information that points to the same URL. In the above example the Actor has url : "https://littr.me/~marius" and the webfinger response has a profile link to the same: 
```
{
            "rel": "https://webfinger.net/rel/profile-page",
            "type": "text/html",
            "href": "https://littr.me/~marius"
}
``` 

Is adding this kind of verification logic feasible or desired?

**Comment by guysoft** [comment]
So I am writing a federated group system.
I am trying to get the basic webfinger and profile page right.
it federates fine on Pleroma but not on Mastodon.

Source is here:
https://github.com/guysoft/fedigroup/blob/main/src/app/main.py#L207

Here is the debug output from mastodon for a non-successful lookup of the user:

This is the one working (pleroma):
https://pastebin.com/VUaKuKCK

This is the one I wrote failing (fedigroup):
https://pastebin.com/E413NaUm

i can also see on the app debug side that mastodon reads the webfinger, but not the profile page, so it stops halfway

**Comment by guysoft** [comment]
Mastodon search API is really just using webfinger, then looking out the actor profile of the user. What my output shows is that the fronted is unnecessary to reproduce the issue. Just the webfinger (and seeing it does not look up the actor profile).

If you also provide the mastodon log with ``RAILS_LOG_LEVEL=debug`` we can confirm where Mastodon gets stuck and it means we  its simple enough that I can rip out that code and have a few lines reproduction script, something anyone can run here. Thats something we could figure how to fix this with.

**Comment by mariusor** [comment]
Maybe I didn't explain myself properly in my comment: your issue is different than my issue.

By using this ticket for your problem you're polluting the discussion and make it harder for eventual contributors to fix either of them. 

I would appreciate it if you could provide proper data if your problem actually occurs because of multiple webfinger requests going from host to host, or moving your comments to a different ticket.

**Comment by guysoft** [comment]
Made clear, I will open  seprate issue and put my findings there.

If its ok with you I will mention this issue there.

**Comment by MFTabriz** [comment]
> As Mastodon requires _all_ actors it interacts with to have a canonical `acct:` URI, it must be able to find that canonical `acct:` from the actor's `id`. This is done by fetching the actor, using `preferredName`, and doing one or more webfinger requests. **In this case, to be compatible with Mastodon, `https://federated.id/.well-known/webfinger?resource=acct:marius@federated.id` should redirect to (or return the same contents as) `https://littr.me/.well-known/webfinger?resource=acct:marius@littr.me`**

Does this mean the response to `https://federated.id/.well-known/webfinger?resource=acct:marius@federated.id` query should have the `subject` set as `acct:marius@federated.id` or `acct:marius@littr.me`?

**Comment by mariusor** [comment]
I'm not sure if this question is addressed to me, but I'll try to offer my perspective:

For all intents and purposes both resources `acct:marius@littr.me`, `acct:marius@federated.id` refer to the same ActivityPub actor at: https://federated.id/actors/8b740680-ccb4-4265-82e7-4ac2ca402750

The resources as expressed in `acct:preferredUsername@instance.domain` are just conveniences that Mastodon created for making user discovery possible, and from the point of view of the ActivityPub specification don't exist as data associated with the Actor.

This ticket is about finding a compromise so not all nodes require implementing the webfinger part (in this specific case, the pure ActivityPub server federated.id does not have it) but still allow user discovery in Mastodon.

So, after Mastodon receives a webfinger response from the littr.me server that gives it an Actor URL from a different domain we should find a way of not requiring another webfinger call to that domain for the resource.

Please let me know if I managed to answer the question, or there should be something more to address.

**Comment by greg-kennedy** [comment]
Spent three hours trying to figure out why my Follow request couldn't be accepted by my homegrown service, turns out it's because the ActivityPub spec says `preferredUsername` is _optional_ ("MAY") but it is evidently required for Mastodon or else it makes a bogus WebFinger request to acct:@domain (instead of acct:user@domain).

Maybe the spec should be updated.

**Comment by mariusor** [comment]
@greg-kennedy I see it the other way: Mastodon should account for `preferredUsername` being missing and fall back to `Name` and only if that also being missing ignore the actor all together.

**Comment by ClearlyClaire** [comment]
> Spent three hours trying to figure out why my Follow request couldn't be accepted by my homegrown service, turns out it's because the ActivityPub spec (4.1 Actor Objects) says `preferredUsername` is _optional_ ("MAY") but it is evidently required for Mastodon or else it makes a bogus WebFinger request to `acct:@domain` (instead of `acct:user@domain`, or more correctly: `user` is blank).
> 
> Maybe the spec should be updated.

ActivityPub is not a self-sufficient spec, and Mastodon has extra requirements (that are, at least in the Webfinger case, documented in `FEDERATION.md` and https://docs.joinmastodon.org/spec/). Furthermore, when rejecting a signed query, Mastodon should return a short error text along with the 401 or 403 status code.

That being said, Mastodon should probably error out with a specific error instead of trying to query `acct:@domain`.

**Comment by greg-kennedy** [comment]
Both of those seem like good solutions... I know I tried `name` before stumbling on `preferredUsername`, it seemed more intuitive to me.  And a descriptive response from the server, like "Could not find `name` in Actor json" would've saved a LOT of time.  Most of the other error messages directed me the right way to fixing bugs, like warnings about wrong URLs etc.

**Comment by ClearlyClaire** [comment]
Tried to add a few more specific error messages in #21617 to help with debugging
